ddl/column_type_change_test.go

```diff
@@ -1167,3 +1167,377 @@ func (s *testColumnTypeChangeSuite) TestColumnTypeChangeFromDateTimeTypeToOthers
 	tk.MustExec("alter table t modify y json")
 	tk.MustQuery("select * from t").Check(testkit.Rows("\"2020-10-30\" \"19:38:25.001\" \"2020-10-30 08:21:33.455555\" \"2020-10-30 08:21:33.455555\" 2020"))
 }
+
+func (s *testColumnTypeChangeSuite) TestColumnTypeChangeFromJsonToOthers(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+
+	// Set time zone to UTC.
+	originalTz := tk.Se.GetSessionVars().TimeZone
+	tk.Se.GetSessionVars().TimeZone = time.UTC
+	defer func() {
+		tk.Se.GetSessionVars().EnableChangeColumnType = false
+		tk.Se.GetSessionVars().TimeZone = originalTz
+	}()
+
+	// Init string date type table.
+	reset := func(tk *testkit.TestKit) {
+		tk.MustExec("drop table if exists t")
+		tk.MustExec(`
+			create table t (
+				obj json,
+				arr json,
+				nil json,
+				t json,
+				f json,
+				i json,
+				ui json,
+				f64 json,
+				str json
+			)
+		`)
+	}
+
+	// To numeric data types.
+	// tinyint
+	reset(tk)
+	tk.MustExec("insert into t values ('{\"obj\": 100}', '[-1, 0, 1]', 'null', 'true', 'false', '-22', '22', '323232323.3232323232', '\"json string\"')")
+	// MySQL will get "ERROR 1366 (HY000) Incorrect integer value: '{"obj": 100}' for column 'obj' at row 1".
+	tk.MustExec("alter table t modify obj tinyint")
+	// MySQL will get "ERROR 1366 (HY000) Incorrect integer value: '[-1, 0, 1]' for column 'arr' at row 1".
+	tk.MustExec("alter table t modify arr tinyint")
```

Comment:
is this expected?

Reply:
I'll try to change it to some similar error.

Reply:
I've made some attempts to try return an error when convert JSON object/array to integer/float. It's work well.

However, there's some related changes need to be done including change tikv push down conversion, fix common test and copr test, etc. 

I think maybe I'll leave the work of refine JSON convert to int to an individual PR with a new issue.

Reply:
The issue: https://github.com/pingcap/tidb/issues/21362